### DISCUSSION TITLE
Most efficient way to draw (Video)frames

### DISCUSSION THREAD

**Discussion by Sekoree** [AvaloniaUI/Avalonia#17013] [Q&A]
Hi, 
I've been trying to make Video control based on FFMPEG, which worked for the most part, at least the basics.
https://github.com/vocawaves/Manager/tree/master/FV.Test

Now my problem, I'm currently drawing all video frames via a WritableBitmap in a Control derived from Image, which works but for 1080p 60fps I get a GPU load between 10-30% (6600XT) depending on the video format, though it also feels a bit random sometimes.

Is there a better way to do this? I was thinking of ICustomDrawOperation but so far all my attempts failed with the Render method never getting called.

**Comment by maxkatz6** [comment]
You can follow this thread for some ideas https://github.com/AvaloniaUI/Avalonia/discussions/16836

**Reply by maxkatz6** [reply]
Also https://github.com/AvaloniaUI/Avalonia/discussions/16997

**Comment by stevemonaco** [comment]
>Is there a better way to do this? I was thinking of ICustomDrawOperation but so far all my attempts failed with the Render method never getting called.

Make sure the `ICustomDrawOperation` has the `Bounds` defined. Ideally, you want to work with composition / gpu interop, but you're deep into territory with little documentation, examples, and communicated design intent. Probably pretty close to platform-dependent code, too.

**Comment by Coloryr** [comment]
![image](https://github.com/user-attachments/assets/af0a7803-951b-4cac-a55a-b6a17061bb1e)
10% with 4070S
https://github.com/Coloryr/FFClient

**Reply by Coloryr** [reply]
```
private void VideoLoad()
{
    _bitmap?.Dispose();
    _bitmap = new(new PixelSize(_handel!.Width, _handel.Height), new Vector(96, 96),
        PixelFormat.Bgra8888, AlphaFormat.Opaque);
    
    Dispatcher.UIThread.Post(() =>
    {
        Image1.Source = _bitmap;
        level?.RequestAnimationFrame((t) =>
        {
            Render();
        });
    });
}

private void Render()
{
    
    unsafe
    {
        using var locked = _bitmap!.Lock();
        Unsafe.CopyBlock(locked.Address.ToPointer(), _handel!.Ptr.ToPointer(),
                (uint)(_handel.Width * _handel.Height * 4));
    }

    Image1.InvalidateVisual();

    level?.RequestAnimationFrame((t) =>
    {
        Render();
    });
}
```

**Reply by Sekoree** [reply]
Thank you, that got me to 60fps on up to 2 windows when sharing the same WriteableBitmap, after that (depending on Video) it seems to lock itself to ~30fps. 
Scaling the image seems to take up most of the resources, so maybe some GPU interop or using Skia directly might be better in the future

**Reply by Coloryr** [reply]
`WriteableBitmap` is just the `SKBitmap` wrapper, copy is necessary
